provider/consul_catalog.go

```diff
@@ -0,0 +1,222 @@
+
+			data, meta, err := catalog.Services(opts)
+			if err != nil {
+				log.WithError(err).Errorf("Failed to list services")
+				return
+			}
+
+			// If LastIndex didn't change then it means `Get` returned
+			// because of the WaitTime and the key didn't changed.
+			if opts.WaitIndex == meta.LastIndex {
+				continue
+			}
+			opts.WaitIndex = meta.LastIndex
+
+			if data != nil {
+				watchCh <- data
+			}
+		}
+	}()
+
+	return watchCh
+}
+
+func (provider *ConsulCatalog) watchService(service string, stopCh <-chan struct{}, services chan<- catalogUpdate) {
+	catalog := provider.client.Catalog()
+
+	go func() {
+		opts := &api.QueryOptions{WaitTime: DefaultWatchWaitTime}
+
+		select {
+		case <-stopCh:
+			return
+		default:
+		}
+
+		data, _, err := catalog.Service(service, "", opts)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to fetch details of " + service)
+			return
+		}
+
+		if data != nil {
+			services <- catalogUpdate{
+				Service: service,
+				Nodes:   data,
+			}
+		}
+	}()
+}
+
+func (provider *ConsulCatalog) getFrontendValue(service string) string {
+	return service + "." + provider.Domain
+}
+
+func (provider *ConsulCatalog) buildConfig(catalog map[string][]*api.CatalogService) *types.Configuration {
+	var FuncMap = template.FuncMap{
+		"getFrontendValue": provider.getFrontendValue,
+		"replace":          replace,
+	}
+
+	allNodes := []*api.CatalogService{}
+	serviceNames := []string{}
+	for service, nodes := range catalog {
+		if len(nodes) > 0 {
+			serviceNames = append(serviceNames, service)
+			allNodes = append(allNodes, nodes...)
+		}
+	}
+
+	templateObjects := struct {
+		Services []string
+		Nodes    []*api.CatalogService
+	}{
+		Services: serviceNames,
+		Nodes:    allNodes,
+	}
+
+	configuration, err := provider.getConfiguration("templates/consul_catalog.tmpl", FuncMap, templateObjects)
+	if err != nil {
+		log.WithError(err).Error("Failed to create config")
+	}
+
+	return configuration
+}
+
+func (provider *ConsulCatalog) watch(configurationChan chan<- types.ConfigMessage) error {
+	var servicesStopCh chan struct{}
+	var pending map[string]bool
+	stopCh := make(chan struct{})
+	services := make(chan catalogUpdate)
+	serviceCatalog := provider.watchServices(stopCh)
+
+	nodes := make(map[string][]*api.CatalogService)
+
+	defer func() {
+		if servicesStopCh != nil {
+			close(servicesStopCh)
+		}
+		close(stopCh)
+	}()
+
+	for {
+		select {
+		case index, ok := <-serviceCatalog:
+			if !ok {
+				return errors.New("Consul service list nil")
+			}
+			log.Debug("List of services changed")
+			if servicesStopCh != nil {
+				close(servicesStopCh)
+			}
+			servicesStopCh = make(chan struct{})
+			pending = make(map[string]bool)
+			for service := range index {
+				name := strings.ToLower(service)
+				if !strings.Contains(name, " ") && !pending[name] {
+					log.WithFields(log.Fields{
+						"service": name,
+					}).Debug("Fetching service")
+					provider.watchService(name, servicesStopCh, services)
```

Comment:
This is not actually a watch any more but I forgot to rename the function. It will however still create 300 goroutines in that scenario when a change happens to fetch the nodes of each service but each goroutine only does one request and then exits.